### ISSUE TITLE
Help with simple workflow in Elsa 2.0

### ISSUE THREAD

**Issue by tomy2105** [elsa-workflows/elsa-core#769]
Hi everyone, need some help.

Was toying with Elsa 1.5 and manged to get designer to write simple workflow with http endpoing and write response displaying only the method in the response content.

Wanted to try Elsa 2.0, since you are doing great stuff in it :), and managed to compile from source (after few trials and erros) and am writing same simple workflow in new designer.

However I cannot get variable to work in response (hardcoded stuff works).
I'm probably missing something that needs to be done with new checkboxes about saving/loading outputs, context, etc.....
Can someone please help me and point me in the right direction.

I seem to have problems attaching here directly so the link to my simple workflow is here:
[https://www.dropbox.com/s/iwwhlo5uqjvdrjm/http_endpoint_with_response.json?dl=0](url)

Thank you in advance :)

**Comment by sfmskywalker** [comment]
I downloaded your JSON (thanks for that!) and imported it into the editor.
What I see is the following JS expression in the **HTTP response** activity:

`Listen.Method`

I'm not sure what you are intending to do, but AFAIK `Listen` is undefined.
If you simply want to test JS evaluation, perhaps try this:

1. Update the JS expression with this: ``Hello, ${queryString('name')}!``
2. Open a webbrowser and go to: `http://localhost:5000/test?name=tomy2015` (make sure you use the appropriate port number)

You should now see a response with:

`Hello, tomy2015!`

Here's an animated gif showing what I did + result:

![request-response-js](https://user-images.githubusercontent.com/938393/111636565-e5aea800-87f8-11eb-9d8c-4b3d64ed8041.gif)

**Comment by tomy2105** [comment]
Thanks for an speedy response.
In 1.5 I used lastResult to extract result of HTTP endpoint and it returned object that had members Path, QueryString, Headers, Method and Body (if read content was enabled). And I used those.

Now in 2.0, I couldn't find lastResult anymore. So I filled Name of the HTTP Endpoint activity to be "Listen" and tried to use it as variable, but that didn't work either......

So my question is how to I get result of HTTP endpoint (Path, QueryString, Headers, Method and Body) stored in variable so I can use them in (Javascript) expressions?
You showed me to use queryString, which is fine and works (though specific to HTTP endpoint activity) , but how do I get HTTP headers, or body of the HTTP request, or the path it was executed against?
Generic question how do I store "result" of any activity and access it from Javascript expressions?

**Comment by tomy2105** [comment]
In other words, how would I write in designer following workflow (from your sample :))?
How do I, from designer, access what you do below with ((HttpRequestModel)(context.Input))?.Body)?

    public class RegistrationWorkflow : IWorkflow
    {
        public void Build(IWorkflowBuilder builder)
        {
            builder
                // Configure a Receive HTTP Request trigger that executes on incoming HTTP POST requests.
                .HttpEndpoint(activity => activity.WithPath("/register").WithMethod(HttpMethods.Post).WithTargetType<Registration>())
                
                // Store the registration as a workflow variable for easier access.
                .SetVariable(context => (Registration)((HttpRequestModel)(context.Input))?.Body)
                
                // Correlate the workflow by email address.
                .Correlate(context => context.GetVariable<Registration>()!.Email)

                // Write an HTTP response with a hyperlink to continue the workflow (notice the presence of the "correlation" query string parameter). 
                .WriteHttpResponse(activity => activity
                    .WithStatusCode(HttpStatusCode.OK)
                    .WithContentType("text/html")
                    .WithContent(context =>
                    {
                        var registration = context.GetVariable<Registration>()!;
                        return $"Welcome onboard, {registration.Name}! Please <a href=\"http://localhost:8201/confirm?correlation={registration.Email}\">confirm your registration</a>";
                    }))
                
}

**Comment by sfmskywalker** [comment]
Got it.

In the designer, when using the `SetVariable` activity (directly connected to `HttpEndpoint`), you could use the following JS expression to store the received `HttpRequestModel`:

```javascript
input
```

The `input` variable is set by Elsa to the value of the `Input` property on the activity execution context, and is always the same value that was returned as output from the previously executed activity. So, since `HttpEndpoint` returns an output of type `HttpRequestModel`, you will get that output as an input into whatever activity executes next (e.g. SetVariable).

The `HttpRequestModel` class looks like this:

```csharp
public Uri Path { get; set; } = default!;
public string Method { get; set; } = default!;
public IDictionary<string, StringValuesModel> QueryString { get; set; } = default!;
public IDictionary<string, StringValuesModel> Headers { get; set; } = default!;
public object? Body { get; set; }
public T GetBody<T>() => (T)Body!;
```

You should be able to access each member from JS expressions. And to access fields on the `Body` object, you should be able to use dot-notation for that. If not `TargetType` value was set on the `HttpEndpoint`, the type for `Body` will be `ExpandoObject`, which allows you to access JSON payloads using dot-notation as well.

For example, if you post the following JSON payload:

```json
{
   "Author": "John"
}
```

Then you should be able to write back the author name from `WriteHttpResponse` with the following JS expression:

```javascript
`The name of the author is ${input.Body.Author}`
```

**Comment by sfmskywalker** [comment]
Here's a working example of accessing the posted JSON payload and writing back the author name:

[workflow-definition-request-response.zip](https://github.com/elsa-workflows/elsa-core/files/6170672/workflow-definition-request-response.zip)

**Comment by sfmskywalker** [comment]
And here's a working example (after you pull latest from the `feature/elsa-2.0` branch because it wasn't working as advertised 😅 ) that demonstrates storing the HTTP Request model into a variable, which is then used in the HTTP Response activity:

![image](https://user-images.githubusercontent.com/938393/111772722-7c3ca100-88ad-11eb-8844-6db85002c8a2.png)

[workflow-definition-request-variable-response.zip](https://github.com/elsa-workflows/elsa-core/files/6170700/workflow-definition-request-variable-response.zip)

**Comment by sfmskywalker** [comment]
And for completeness' sake, the following JS syntax can be used to access a given activity's output directly:

```javascript
`The name of the author is ${Listen.Output.Body.Author}`
```

Where "Listen" is the name of the `HttpEndpoint` activity.

![image](https://user-images.githubusercontent.com/938393/111773431-74c9c780-88ae-11eb-9580-ca2311349e18.png)

[workflow-definition-request-variable-response-using-named-activity.zip](https://github.com/elsa-workflows/elsa-core/files/6170749/workflow-definition-request-variable-response-using-named-activity.zip)

**Comment by tomy2105** [comment]
Thank you again for the quick response, awesome work!
Is there (sorry if I missed it), or will there be, documentation page which describes "input" variable as well as models of all the activities?
It is not too important for me (can always look at the code) but for people only wanting to use designer it might be beneficial.

One other question.
If you need to do Listen.Output.Method, is there anything else in Listen object besides Output? Will there be more here in future?

Thanks for all the help!

**Comment by sfmskywalker** [comment]
> Is there (sorry if I missed it), or will there be, documentation page which describes "input" variable as well as models of all the activities?

There's no documentation for Elsa 2 at the moment, but we will be working on that soon as we got the designer bits ironed out and are ready to release 2.0. The documentation will describe each activity, including their output models.

> If you need to do Listen.Output.Method, is there anything else in Listen object besides Output? Will there be more here in future?

Great question! I haven't figured this one out yet, but my current thinking is to allow for this to be extensible in the future. So to be safe, I added the `Output` property. Perhaps in the future we'll allow you to access the evaluated property values. But this would require the activity to actually persist these values. Which of course could be done by the runtime, but that also means the size of a workflow instance might grow. So perhaps this is something we want the user to be in control of. 

As I said: something to figure out still 😅
